gk/fib.c

```diff
@@ -0,0 +1,1440 @@
+/*
+ * Gatekeeper - DoS protection system.
+ * Copyright (C) 2016 Digirati LTDA.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "gatekeeper_fib.h"
+#include "gatekeeper_gk.h"
+#include "gatekeeper_main.h"
+
+struct ip_prefix {
+	const char    *str;
+	struct ipaddr addr;
+	int           len;
+};
+
+void
+destroy_neigh_hash_table(struct neighbor_hash_table *neigh)
+{
+	if (neigh->cache_tbl != NULL) {
+		rte_free(neigh->cache_tbl);
+		neigh->cache_tbl = NULL;
+	}
+
+	if (neigh->hash_table != NULL) {
+		rte_hash_free(neigh->hash_table);
+		neigh->hash_table = NULL;
+	}
+}
+
+static inline int
+gk_lpm_add_ipv4_route(uint32_t ip,
+	uint8_t depth, uint32_t nexthop, struct gk_lpm *ltbl)
+{
+	return rte_lpm_add(ltbl->lpm, ntohl(ip), depth, nexthop);
+}
+
+static inline int
+gk_lpm_add_ipv6_route(uint8_t *ip,
+	uint8_t depth, uint32_t nexthop, struct gk_lpm *ltbl)
+{
+	return rte_lpm6_add(ltbl->lpm6, ip, depth, nexthop);
+}
+
+static inline int
+initialize_ether_cache(struct ether_cache *eth_cache)
+{
+	/*
+	 * TODO Add concurrency control to
+	 * reset the Ethernet cache entry.
+	 */
```

Comment:
`initialize_ether_cache()` is only called on cache entries that are not being used, so we don't need a concurrencty mechanism here. You should drop the comment above and add a comment to `initialize_ether_cache()` to explain that callers must ensure that the entry is not being used.